Enable and apply flake8-comprehensions linter rules

Skylion007 opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->
Fixes #24740 

#### Brief description of what is fixed or changed
* This enable flake8-comprehensions plugin which is a linter that helps write more performant collection comprehensions in Python. All the rules are designed in a way that optimizes performance and the generated bytecode in Python.
* Almost all fixes are automatically generated by the ruff linters autofixes for the corresponding flake8 plugin so while this is a massive PR, almost all fixes are automated and therefore should be safe. All tests are passing too which validates this.

#### Other comments


#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers. Formerly, `log(-x)` incorrectly gave `-log(x)`.

* physics.units
  * Corrected a semantical error in the conversion between volt and statvolt which
    reported the volt as being larger than the statvolt.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
NO ENTRY
<!-- END RELEASE NOTES -->

Skylion007 committed:
Enable and apply flake8-comprehensions linter rules

Skylion007 committed:
Make config parser happy

Skylion007 committed:
Apply a few manual fixes

Skylion007 committed:
A few more manual fixes

Skylion007 committed:
Update .mailmap

Skylion007 committed:
Revert change in excluded file

oscarbenjamin commented on sympy/assumptions/ask_generated.py:
```diff
@@ -97,168 +97,168 @@ def get_known_facts_dict():
-        Q.complex_elements: (set([Q.complex_elements]), set([])),
-        Q.composite: (set([Q.algebraic, Q.commutative, Q.complex, Q.composite,
+        Q.algebraic: ({Q.algebraic, Q.commutative, Q.complex, Q.finite},
+        {Q.infinite, Q.negative_infinite, Q.positive_infinite,
```
This file is automatically generated and should not be modified without modifying the script that generates it.

Skylion007 replied:
Ah, I'll fix that. @oscarbenjamin Any other autogenerated files I should be aware of?

oscarbenjamin commented on sympy/codegen/numpy_nodes.py:
```diff
@@ -60,7 +60,7 @@ def _eval_evalf(self, *args, **kwargs):
 
     def _eval_simplify(self, *args, **kwargs):
-        a, b = map(lambda x: x.simplify(**kwargs), self.args)
+        a, b = (x.simplify(**kwargs) for x in self.args)
```
Side comment:

I wish we wouldn't have two element loops like this. This line would fail if there are not two args so just write the signature like that:
```python
def _eval_simplify(self, a, b, **kwargs):
    a = a.simplify(**kwargs)
    b = b.simplify(**kwargs)
```
The repetition is fine: it is not worth complicating things just to reduce a single simple duplicate call.

Skylion007 replied:
@oscarbenjamin Actually, make this change apparently broke tests. https://github.com/sympy/sympy/actions/runs/4214044376/jobs/7314304910 reverting for now

oscarbenjamin replied:
How can that break anything? Is something catching an exception from this somehow?

Skylion007 replied:
Ah, @oscarbenjamin `self.args != args`

oscarbenjamin replied:
Ah okay.

Skylion007 committed:
Do no modify generated file. Add exclusion

oscarbenjamin commented on sympy/core/assumptions_generated.py:
```diff
@@ -37,42 +37,33 @@
-full_implications = dict( [
-    # Implications of algebraic = True:
-    (('algebraic', True), set( (
+full_implications = { ('algebraic', True): {
```
This file is also automatically generated. It should only be changed by the script that generates it.

Skylion007 committed:
Revert another generated file. Add another exclusion

oscarbenjamin commented on sympy/parsing/autolev/_antlr/autolevparser.py:
```diff
@@ -3029,7 +3029,7 @@ def expr(self, _p:int=0):
     def sempred(self, localctx:RuleContext, ruleIndex:int, predIndex:int):
         if self._predicates == None:
-            self._predicates = dict()
+            self._predicates = {}
```
Also autogenerated

oscarbenjamin commented on sympy/parsing/latex/_antlr/latexparser.py:
```diff
@@ -3604,7 +3604,7 @@ def supeq(self):
     def sempred(self, localctx:RuleContext, ruleIndex:int, predIndex:int):
         if self._predicates == None:
-            self._predicates = dict()
+            self._predicates = {}
```
Autogenerated

Skylion007 committed:
One more autogenerated file

Skylion007 committed:
One more autogenerated file

Skylion007 committed:
Reorder exclusions

oscarbenjamin commented on sympy/polys/specialpolys.py:
```diff
@@ -180,7 +180,7 @@ def fateman_poly_F_1(n):
     y_0, y_1 = Y[0], Y[1]
 
-    u = y_0 + Add(*[y for y in Y[1:]])
+    u = y_0 + Add(*list(Y[1:]))
```
There's no point in having list here. It can just be `Add(*Y[1:])`. Same below.

Skylion007 committed:
Apply reviewer comments

oscarbenjamin commented on sympy/printing/fortran.py:
```diff
@@ -387,8 +387,7 @@ def _print_AugmentedAssignment(self, expr):
         return self._get_statement("{0} = {0} {1} {2}".format(
-            *map(lambda arg: self._print(arg),
-                 [lhs_code, expr.binop, rhs_code])))
+            *(self._print(arg) for arg in [lhs_code, expr.binop, rhs_code])))
```
I don't think that this 3 element loop is worth it either. Better to make is explicit:
```
"{0} = {0} {1} {2}".format(self._print(lhs_code), self._print(expr.binop), self._print(rhs_code))
```
Otherwise I would use `map` but the `lambda` here is unnecessary:
```
"{0} = {0} {1} {2}".format(map(self._print, [lhs_code, expr.binop, rhs_code]))
```

oscarbenjamin commented on sympy/sets/sets.py:
```diff
@@ -2600,7 +2600,7 @@ def simplify_intersection(args):
                 return Union(*(Intersection(arg, other) for arg in s.args))
             else:
-                return Union(*[arg for arg in s.args])
+                return Union(*list(s.args))
```
Here `list` is unnecessary.

oscarbenjamin commented on sympy/stats/rv.py:
```diff
@@ -1446,7 +1446,7 @@ def sampling_E(expr, given_condition=None, library='scipy', numsamples=1,
     samples = list(sample_iter(expr, given_condition, library=library,
                           numsamples=numsamples, seed=seed, **kwargs))
-    result = Add(*[samp for samp in samples]) / numsamples
+    result = Add(*list(samples)) / numsamples
```
List unnecessary

oscarbenjamin commented on sympy/stats/stochastic_process_types.py:
```diff
@@ -485,7 +485,7 @@ def _work_out_state_index(self, state_index, given_condition, trans_probs):
             raise ValueError("state space is not compatible with the transition probabilities.")
         if not isinstance(trans_probs.shape[0], Symbol):
-            state_index = FiniteSet(*[i for i in range(trans_probs.shape[0])])
+            state_index = FiniteSet(*list(range(trans_probs.shape[0])))
```
List unnecessary

Skylion007 committed:
More reviewer comments

oscarbenjamin commented on sympy/stats/stochastic_process_types.py:
```diff
@@ -1256,7 +1256,7 @@ def stationary_distribution(self, condition_set=False) -> tUnion[ImmutableMatrix
 
         soln = list(linsolve((a, b)))[0]
-        return ImmutableMatrix([[sol for sol in soln]])
+        return ImmutableMatrix([list(soln)])
```
List unnecessary

oscarbenjamin commented on sympy/stats/stochastic_process_types.py:
```diff
@@ -1612,7 +1612,7 @@ def limiting_distribution(self):
         eqs.append(sum(wi) - 1)
         soln = list(linsolve(eqs, wi))[0]
-        return ImmutableMatrix([[sol for sol in soln]])
+        return ImmutableMatrix([list(soln)])
```
List unnecessary

Skylion007 committed:
More reviewer comments

oscarbenjamin commented on .github/workflows/runtests.yml:
```diff
@@ -22,7 +22,7 @@ jobs:
           python-version: '3.10'
       - run: python -m pip install --upgrade pip
-      - run: pip install mpmath flake8
+      - run: pip install mpmath flake8 flake8-comprehensions
```
Does installing this cause it to be run automatically by flake8?

Skylion007 replied:
@oscarbenjamin No, the rules have to be manually selected in the setup.cfg

oscarbenjamin replied:
Is it because you've added the C code in setup.cfg?

https://github.com/sympy/sympy/blob/7ba640371de813a3c59b46d862bb019d21ac78ee/setup.cfg#L5-L15

Skylion007 replied:
Yes, all the error codes begin with C4: https://pypi.org/project/flake8-comprehensions/

oscarbenjamin replied:
So if flake8-comprehensions is installed and the C code is added in setup.cfg then that causes these checks to run?

I just want to be clear so that anyone looking at this in future can understand.

Also you have added other files to the exclude list here because they are automatically generated and so probably should not be inspected by these tools.

Skylion007 replied:
Correct, if the plugin is installed it will add the C4 codes to flake8, which can be selected like other flake8 rules.

Yeah, I added the other generated files to the list of ignores since flake8 shouldn't complain about them.

Skylion007 replied:
I originally didn't bother enabling C417 because it didn't have autofixits, but I've went and manually fixed it so all the comprehension checks should be enabled.

Skylion007 committed:
Enable rule C417 too

oscarbenjamin commented:
Okay, I just skimmed back through very quickly but I think it looks good.

Thanks @Skylion007!

Skylion007 commented:
> Benchmark results from GitHub Actions
> 
> Lower numbers are good, higher numbers are bad. A ratio less than 1 means a speed up and greater than 1 means a slowdown. Green lines beginning with `+` are slowdowns (the PR is slower then master or master is slower than the previous release). Red lines beginning with `-` are speedups.
> 
> Significantly changed benchmark results (PR vs master)
> 
> Significantly changed benchmark results (master vs previous release)
> 
> ```diff
>        before           after         ratio
>      [41d90958]       [f554606e]
>      <sympy-1.11.1^0>                 
> -       989±0.6μs          580±1μs     0.59  solve.TimeSparseSystem.time_linear_eq_to_matrix(10)
> -     2.81±0.01ms      1.08±0.01ms     0.38  solve.TimeSparseSystem.time_linear_eq_to_matrix(20)
> -     5.60±0.01ms         1.58±0ms     0.28  solve.TimeSparseSystem.time_linear_eq_to_matrix(30)
> ```
> 
> Full benchmark results can be found as artifacts in GitHub Actions (click on checks at the top of the PR).

Neat, I wasn't expecting massive speed up from this PR, but looks like it gave at least a 4X speed up impressively.

Skylion007 committed:
Revert reviewer change

oscarbenjamin commented:
> Neat, I wasn't expecting massive speed up from this PR, but looks like it gave at least a 4X speed up on at least 1 codepath impressively.

This is a nice PR but more from a stylistic perspective. Those benchmark results are actually referring to cumulative changes since the last release of SymPy ("master vs previous release").

Skylion007 commented:
> > Neat, I wasn't expecting massive speed up from this PR, but looks like it gave at least a 4X speed up on at least 1 codepath impressively.
> 
> This is a nice PR but more from a stylistic perspective. Those benchmark results are actually referring to cumulative changes since the last release of SymPy ("master vs previous release").

Yeah, I realized my error. Hence, I deleted my comment. :P

brocksam commented:
If PR #24729 is merged in addition to this we'd also be easily able to enforce the C4 rules with Ruff too.

sylee957 commented:
I guess this is okay to merge. But often the changes in CI are not reflected in other ongoing PRs so we may need to wait for other people to validate their own work against flake8 comprehensions.

hanspi42 commented:
Maybe one thing should be changed:

> So if flake8-comprehensions is installed and the C code is added in setup.cfg then that causes these checks to run?
> 
> I just want to be clear so that anyone looking at this in future can understand.

I would only add C4, not C, which is what flake8-comprehensions recommends.  While it does not have anything except C4 right now, we do not know what will happen in the future.

hanspi42 commented on setup.cfg:
```diff
@@ -5,10 +5,12 @@ long_description_content_type = text/markdown
 ignore = F403
-select = F,E722
-exclude = sympy/parsing/latex/_antlr/*,
+select = C,F,E722
```
Only select C4, not C

Skylion007 committed:
Select only C4 error codes

hanspi42 commented:
We now have three people giving the OK; I enable auto-merge to merge this as soon as the checks have completed. Thanks for your contribution!
